### ISSUE TITLE
Pyright does not respect `_generate_next_value_` overrides in enums

### ISSUE THREAD

**Issue by Ovsyanka83** [microsoft/pyright#3742]
When `_generate_next_value_` is overriden, mypy will mark auto() enum values as the overriden type while pyright will still mark them as ints.

**To Reproduce**
python 3.11.0b4 (same behavior on 3.10.4)

```python
from enum import Enum, auto


class StrEnum(str, Enum):
    def _generate_next_value_(name, start, count, last_values) -> str:
        return name


class B(StrEnum):
    c = auto()


reveal_type(B.c.value)

```

**Expected behavior**
The actual value of `B.c.value` is going to be `str` in the scenario above so pyright must reveal the type to be `str`.

**VS Code extension or command-line**
Pylance vscode extension `v2022.7.40`

**Comment by erictraut** [comment]
The `enum` class has a bunch of non-standard behaviors that cannot be described in the Python type system. I recommend sticking to more typical usage patterns if you want to use enums it with static type checking.

We may add support for more of these non-standard capabilities in the future but only if we receive sufficient signal from users that they are commonly used. I'm going to close this feature request for now, but we could revisit it in the future.

**Comment by Ovsyanka83** [comment]
@erictraut the example I provided is taken straight from documentation and I feel like if mypy covers it, then it is a part of standard capabilities. 
![image](https://user-images.githubusercontent.com/17561586/180666412-263dad02-1de7-44b0-acf2-602a50b38bc5.png)

**Comment by erictraut** [comment]
Yes, I understand that the behavior is documented. There are many behaviors of the `Enum` class that are documented but still differ from all other classes in Python. These behaviors are not (and cannot be) described using the typeshed-provided type stubs because the behaviors are not possible to capture in the type system. That's the problem with classes that deviate significantly from standard behaviors. The `Enum` class is probably the most egregious example of this in stdlib. In most programming languages, an enumeration is a simple concept. In Python, it's the extremely complicated and extensible beast.

Pyright already contains a bunch of special-case logic to deal with more commonly-used aspects of `Enum`, but there are many other aspects where special-case logic doesn't exist. In general, we don't like adding such special-case logic because it is fragile (since it overrides the information provided in typeshed, and the two can get out of sync), and it increases the maintenance burden of the type checker.

You are the first person in the past 3 years to request this feature for pyright, and it's the first time I've even heard of it. So it can't be all that common.

Like I said, if we receive signal that this is used by more than just a couple of Python developers, we would be willing to revisit this. The most common signal we look at is thumbs-up reactions.

**Comment by Ovsyanka83** [comment]
Then I guess it would make sense to reopen this issue because otherwise nobody will see this issue and hence nobody will put thumbs up.

**Comment by erictraut** [comment]
That's not how I manage issues in the pyright project. I only keep issues open if they are planned. This issue is still searchable, and people can add reactions and additional comments to it still.

**Comment by Ovsyanka83** [comment]
A bit more info on its use-cases:
In Python 3.11, StrEnum was added to the enum module:
![image](https://user-images.githubusercontent.com/17561586/180667279-b8fca283-a371-46d6-86e2-2d0b45ac1703.png)

**Comment by Ovsyanka83** [comment]
Currently, typeshed adds types to enum values by using:
```python
class StrEnum(str, ReprEnum):
        def __new__(cls: type[Self], value: str) -> Self: ...
        _value_: str
        @_magic_enum_attr
        def value(self) -> str: ...
```

where `_magic_enum_attr` is a `types.DynamicClassAttribute`

So there's no issue with getting out of sync with typeshed. Alex Waygood (the maintainer of typeshed) and 10 other python developers have supported this issue as well.

**Comment by erictraut** [comment]
That's enough signal to convince me. This will be implemented in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.264, which I just published. It will also be included in a future release of pylance.
